Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+        stdout->write(pattern_sv.substring_view(0, newline_position + 1));
+        break;
+    }
+    case 'q':
+        cycle_decision = CycleDecision::Quit;
+        break;
+    case 'x':
+        swap(pattern_space, hold_space);
+        break;
+    case '=':
+        stdout->write(String::formatted("{}\n", input.line_number()));
+        break;
+    case '#':
+        break;
+    default:
+        warnln("Command not implemented: {}", command.function);
+        break;
+    }
+
+    return cycle_decision;
+}
+
+static bool run(Vector<InputFile>& inputs, Script& script, bool suppress_default_output)
+{
+    // TODO: verify all commands are valid
+
+    StringBuilder pattern_space;
+    StringBuilder hold_space;
+    auto stdout = Core::File::standard_output();
+
+    // TODO: extend to multiple input files
+    auto& input = inputs[0];
+    // main loop
+    while (input.has_next()) {
+
+        // TODO: "Reading from input shall be skipped if a <newline> was in the pattern space prior to a D command ending the previous cycle"
+        pattern_space.append(input.next());
+        auto is_last_line = !input.has_next();
+
+        // Turn commands on/off depending on selection. We need
+        for (auto& command : script.commands())
+            command.enable_for(pattern_space.string_view(), input.line_number(), is_last_line);
+
+        // Go, go, go!
+        CycleDecision cycle_decision = CycleDecision::None;
+        for (auto& command : script.commands()) {
+            if (!command.is_enabled())
+                continue;
+            auto command_cycle_decision = apply(command, pattern_space, hold_space, input, suppress_default_output);
+            if (command_cycle_decision == CycleDecision::Next || command_cycle_decision == CycleDecision::Quit) {
+                cycle_decision = command_cycle_decision;
+                break;
+            }
+        }
+
+        if (cycle_decision == CycleDecision::Next) {
+            continue;
+        } else if (cycle_decision == CycleDecision::Quit) {
+            break;
+        }
+        if (!suppress_default_output)
+            write_pattern_space(stdout, pattern_space);
+        pattern_space.clear();
+    }
+
+    return true;
+}
+
+int main(int argc, char* argv[])
+{
+    bool suppress_default_output = false;
+    Core::ArgsParser arg_parser;
+    Script script;
+    Vector<const char*> pos_args;
+    arg_parser.set_general_help("The Stream EDitor");
+    arg_parser.add_option(suppress_default_output, "suppress default output", nullptr, 'n');
+    arg_parser.add_option(Core::ArgsParser::Option {
+        .requires_argument = true,
+        .help_string = "A file containing script commands",
+        .short_name = 'f',
+        .value_name = "script-file",
+        .accept_value = [&script](const char* script_file) {
+            auto file = Core::File::open(script_file, Core::OpenMode::ReadOnly);
+            if (file.is_error()) {
+                perror("Failed to open script file");
```

Comment:
I wouldn't trust `perror` to tell you the right error here, `file.error()` already has the correct error string anyway:
```suggestion
                warnln("Failed to open script file: {}", file.error());
```

Reply:
Thanks for pointing out, I think I'm still not terribly familiar with the Serenity libs and their intended usage. Fixed now.